nodejs/image.bzl

```diff
@@ -0,0 +1,134 @@
+
+def repositories():
+  # Call the core "repositories" function to reduce boilerplate.
+  # This is idempotent if folks call it themselves.
+  _repositories()
+
+  excludes = native.existing_rules().keys()
+  if "nodejs_image_base" not in excludes:
+    container_pull(
+      name = "nodejs_image_base",
+      registry = "gcr.io",
+      repository = "google-appengine/base",
+      digest = DIGESTS["latest"],
+    )
+  if "nodejs_debug_image_base" not in excludes:
+    container_pull(
+      name = "nodejs_debug_image_base",
+      registry = "gcr.io",
+      repository = "google-appengine/base",
+      digest = DIGESTS["debug"],
+    )
+
+DEFAULT_BASE = select({
+    "@io_bazel_rules_docker//:fastbuild": "@nodejs_image_base//image",
+    "@io_bazel_rules_docker//:debug": "@nodejs_debug_image_base//image",
+    "@io_bazel_rules_docker//:optimized": "@nodejs_image_base//image",
+    "//conditions:default": "@nodejs_debug_image_base//image",
+})
+
+def _runfiles(dep):
+  return dep.default_runfiles.files + dep.data_runfiles.files + dep.files
+
+def _emptyfiles(dep):
+  return dep.default_runfiles.empty_filenames + dep.data_runfiles.empty_filenames
+
+def _dep_layer_impl(ctx):
+  return dep_layer_impl(ctx, runfiles=_runfiles, emptyfiles=_emptyfiles)
+
+_dep_layer = rule(
+    attrs = _container.image.attrs + {
+        # The base image on which to overlay the dependency layers.
+        "base": attr.label(mandatory = True),
+        # The dependency whose runfiles we're appending.
+        "dep": attr.label(mandatory = True, allow_files = True),
+
+        # Whether to lay out each dependency in a manner that is agnostic
+        # of the binary in which it is participating.  This can increase
+        # sharing of the dependency's layer across images, but requires a
+        # symlink forest in the app layers.
+        "agnostic_dep_layout": attr.bool(default = False),
+        # The binary target for which we are synthesizing an image.
+        # This is needed iff agnostic_dep_layout.
+        "binary": attr.label(mandatory = False),
+
+        # Override the defaults.
+        # https://github.com/bazelbuild/bazel/issues/2176
+        "data_path": attr.string(default = "."),
+        "directory": attr.string(default = "/app"),
+    },
+    executable = True,
+    outputs = _container.image.outputs,
+    implementation = _dep_layer_impl,
+)
+
+load("@build_bazel_rules_nodejs//:defs.bzl", "nodejs_binary")
+
+def nodejs_image(name, base=None, data=[], layers=[],
+                 node_modules="//:node_modules", **kwargs):
+  """Constructs a container image wrapping a nodejs_binary target.
+
+  Args:
+    layers: Augments "deps" with dependencies that should be put into
+           their own layers.
+    **kwargs: See nodejs_binary.
+  """
+  binary_name = name + ".binary"
+
+  layers = [
+    # Put the Node binary into its own layer.
+    "@nodejs//:bin/node",
+    # node_modules can get large, it should be in its own layer.
+    node_modules,
```

Comment:
if you're concerned about size, you could try filtering out stuff.
This isn't the right spot to do it - but where the `node_modules` filegroup was declared, it's nice to drop things like test data that npm authors lazily published along with their package

Reply:
Yes, I think we want this in `rules_nodejs` :)

Reply:
I filter */test/** now, hopefully that helps and doesn't break things